source/extensions/filters/network/common/redis/fault_impl.cc

```diff
@@ -0,0 +1,141 @@
+    break;
+  default:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+    break;
+  }
+
+  // Get the default value/runtime key
+  default_value_ = base_fault.fault_enabled().default_value();
+  runtime_key_ = base_fault.fault_enabled().runtime_key();
+};
+
+std::vector<std::string> FaultManagerImpl::FaultImpl::buildCommands(
+    envoy::extensions::filters::network::redis_proxy::v3::RedisProxy_RedisFault base_fault) {
+  std::vector<std::string> commands;
+  for (const std::string& command : base_fault.commands()) {
+    commands.emplace_back(absl::AsciiStrToLower(command));
+  }
+  return commands;
+}
+
+FaultManagerImpl::FaultManagerImpl(
+    Runtime::RandomGenerator& random, Runtime::Loader& runtime,
+    const Protobuf::RepeatedPtrField<
+        ::envoy::extensions::filters::network::redis_proxy::v3::RedisProxy_RedisFault>
+        faults)
+    : fault_map_(buildFaultMap(faults)), random_(random), runtime_(runtime) {}
+
+FaultMap FaultManagerImpl::buildFaultMap(
+    const Protobuf::RepeatedPtrField<
+        ::envoy::extensions::filters::network::redis_proxy::v3::RedisProxy_RedisFault>
+        faults) {
+  // Next, create the fault map that maps commands to pointers to Fault objects.
+  // Group faults by command
+  FaultMap fault_map;
+  for (auto base_fault : faults) {
+    auto fault_ptr = std::make_shared<FaultImpl>(base_fault);
+    if (!fault_ptr->commands().empty()) {
+      for (const std::string& command : fault_ptr->commands()) {
+        fault_map[command].emplace_back(fault_ptr);
+      }
+    } else {
+      // Generic "ALL" entry in map for faults that map to all keys; also add to each command
+      fault_map[FaultManagerKeyNames::get().AllKey].emplace_back(fault_ptr);
+    }
+  }
+
+  // Add the ALL keys faults to each command too so that we can just query faults by command.
+  // Get all ALL_KEY faults.
+  FaultMap::iterator it_outer = fault_map.find(FaultManagerKeyNames::get().AllKey);
+  if (it_outer != fault_map.end()) {
+    // For each ALL_KEY fault...
+    for (const FaultSharedPtr& fault_ptr : it_outer->second) {
+      // Loop through all unique commands other than ALL_KEY and add the fault.
+      FaultMap::iterator it_inner;
+      for (it_inner = fault_map.begin(); it_inner != fault_map.end(); it_inner++) {
+        std::string command = it_inner->first;
+        if (command != FaultManagerKeyNames::get().AllKey) {
+          fault_map[command].push_back(fault_ptr);
+        }
+      }
+    }
+  }
+  return fault_map;
+}
+
+// Fault checking algorithm:
+//
+// For example, if we have an ERROR fault at 5% for all commands, and a DELAY fault at 10% for GET,
+// if we receive a GET, we want 5% of GETs to get DELAY, and 5% to get ERROR. Thus, we need to
+// amortize the percentages.
+//
+// 0. Get random number.
+// 1. Get faults for given command.s
+// 2. For each fault, calculate the amortized fault injection percentage.
+//
+// Note that we do not check to make sure the probabilities of faults are <= 100%!
```

Comment:
Per my other comment can we actually check this?

Or better, can we just have all faults be serially applied? I think it would be much simpler to think about? I'm a little confused about how this works currently.